added: github-actions docs (#130)
mia-platform/documentation: docs/development_suite/deploy/configure-github-actions.md

---
id: configure-github-actions

> title:  Configure deployment with GitHub

Reviewer: ```suggestion
title:  Configure deployment with GitHub Actions
```

> sidebar_label: Configure deployment with GitHub

Reviewer: ```suggestion
sidebar_label: Configure deployment with GitHub Actions
```

slug: "/development_suite/deploy/"
---

> The Mia-Platform Console can be configured to deploy your project through GitHub.

Reviewer: ```suggestion
The Mia-Platform Console can be configured to deploy your project through GitHub Actions.
```

> In order to enable your project to be deployed using GitHub, you should edit the configuration of the project through the Console CMS 

Reviewer: ```suggestion
To enable your project to be deployed using GitHub Actions, you should edit the configuration of the project through the Console CMS 
```

(if you do not have access to it, ask your instance administrator). In particular, the `environments.deploy` object 
of the configuration should contain the following information:

```json
{
  "deploy": {
    "type": "github-actions",
    "providerId": "your-github-provider-id"
  }
} 
```

:::info

> if you haven't set up a GitHub provider yet, [see this page](/development_suite/set-up-infrastructure/configure-provider.mdx) to learn how to configure it.

Reviewer 2: ```suggestion
If you haven't set up a GitHub provider yet, check out [here](/development_suite/set-up-infrastructure/configure-provider.mdx) how to configure it.
```

> Remember that the Access Token configured to your provider needs to have the `admin` scope of the organization.

Reviewer: Why the admin scope? It is not possible to set less scope?

Author: from the documentation 
> All actions against organization webhooks require the authenticated user to be an admin of the organization being managed. Additionally, OAuth tokens require the admin:org_hook scope.

[Link](https://docs.github.com/en/rest/orgs/webhooks?apiVersion=2022-11-28)

Reviewer: We need to use organization wide webhooks or project level webhooks?

Author: under the deployment_status event i found this
> To subscribe to this event, a GitHub App must have at least read-level access for the "Deployments" repository permission

> In order for GitHub to send webhook payloads, your server needs to be accessible from the Internet.

Reviewer: Why it is needed to send a payload? It's needed to trigger the pipeline, I suppose

Author: I copied the warnings that Github puts for managing webhooks. 
In general, when Github calls you the target url that you set in the hook, it sends you a payload anyway.
For example this is the payload of the deployment_status event https://docs.github.com/webhooks-and-events/webhooks/webhook-events-and-payloads#deployment_status

Reviewer: Instead of "your server", it's better to say "your console instance"

Author: ```suggestion
In order for GitHub to send webhook payloads, your console instance needs to be accessible from the Internet.
```

:::

:::info

> if you haven't set up an environment yet, [see this page](/development_suite/set-up-infrastructure/add-environment.mdx) to learn how to configure it.

Reviewer 2: ```suggestion
If you haven't set up an environment yet, check out [here](/development_suite/set-up-infrastructure/add-environment.mdx) how to configure it.
```

    name: Deploy to ENVIRONMENT_NAME environment
    if: ${{ github.event.deployment.environment }} == ENVIRONMENT_NAME
    concurrency: ENVIRONMENT_NAME
    permissions:
      deployments: write
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v1

      - name: Update deployment status (pending)
        if: success()
        uses: chrnorm/deployment-status@v2
        with:
          token: '${{ github.token }}'
          state: 'pending'
          deployment-id: ${{ github.event.deployment.id }}

      - name: Deploy my app
        run: |
          # add your deployment code here
          echo Deploying...
          sleep 10s
          echo Done!

      - name: Update deployment status (success)
        if: success()
        uses: chrnorm/deployment-status@v2
        with:
          token: '${{ github.token }}'
          state: 'success'
          deployment-id: ${{ github.event.deployment.id }}

      - name: Update deployment status (failure)
        if: failure()
        uses: chrnorm/deployment-status@v2
        with:
          token: '${{ github.token }}'
          state: 'failure'
          deployment-id: ${{ github.event.deployment.id }}
```

> When users deploy for the first time, the Console release automatically creates the environments configured inside the project, 

Reviewer 2: ```suggestion
When users deploy for the first time, the Console automatically creates the environments configured inside the project, 
```

otherwise, you can [follow this tutorial](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#creating-an-environment) 
to set up your environments.

> With the above configuration, when a user presses the `Deploy` button inside the console, the following steps will be executed:

Reviewer 3: ```suggestion
With the above configuration, when a user presses the `Deploy` button inside the Console, the following steps will be executed:
```

> With the above configuration, when a user presses the `Deploy` button inside the Console, the following steps will be executed:

Reviewer 2: ```suggestion
With the above configuration, when a user clicks the `Deploy` button inside the Console, the following steps will be executed:
```

Reviewer 2: **Generic guideline**
Use "Press" when referring to a hardware button and "Click" when referring to an item in the UI

> When users deploy for the first time, the Console release automatically creates the environments configured inside the project, 
> otherwise you can [follow this tutorial](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#creating-an-environment) 
> to set up your environments.
> 
> With the above configuration, when a user presses the `Deploy` button inside the console, the following steps will be executed:
> * The Console creates a new deployments depending on environment chosen on GitHub repository
> * On deployments creation GitHub dispatches a new deployment event and triggers the job specified inside the workflow
> * When the pipeline status of deployments changes, the Console updates to show you its current status

Author: ```suggestion
When users deploy for the first time, the Console release automatically creates the environments configured inside the project, 
otherwise, you can [follow this tutorial](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#creating-an-environment) 
to set up your environments.

With the above configuration, when a user presses the `Deploy` button inside the console, the following steps will be executed:
* The Console creates new deployments depending on the environment chosen on the GitHub repository
* On deployments creation GitHub dispatches a new deployment event and triggers the job specified inside the workflow
* When the pipeline status of deployments changes, the Console updates to show you its current status
```

### GitHub Webhooks

When a project is created, the Console subscribes it automatically to GitHub Webhooks.
These webhooks are important to allow the Console to update the project and see your deployment history.

> Currently, the events subscribed are: 

Reviewer: Those are not the event subsribed, but the events which the Console can handle

> Currently, the events subscribed are: 

Author: Corrected the intended sentence

> Currently, the events that Console need to handle are: 

Reviewer 3: ```suggestion
Currently, the events that the Console needs to handle are: 
```

* [deployment_status](https://docs.github.com/en/webhooks-and-events/webhooks/webhook-events-and-payloads#deployment_status)

### Projects created before v10.8.0

> For projects already created before the specified version, you need to subscribe to provider events, manually.

Reviewer 2: ```suggestion
For projects already created before the specified version, you need to subscribe to provider events manually.
```

> For projects already created before specified version, you need to subscribe to provider events, manually.
> To do this you can use the GitHub interface or the API.
> 
> Here's an example of deployments_status subscription

Author: ```suggestion
For projects already created before the specified version, you need to subscribe to provider events, manually.
To do this you can use the GitHub interface or the API.

Here's an example of a deployments_status subscription
```
